serverlessworkflow/modules/ROOT/pages/testing-and-troubleshooting/mocking-opnapi-services-with-wiremock.adoc

```diff
@@ -1 +1,310 @@
+
+For more information about WireMock, see the {wiremock_url}[WireMock documentation].
+
+== Adding a mocked OpenAPI service to your tests
+
+The following procedure describes how to add WireMock to your Serverless Workflow application:
+
+.Procedure
+
+. Add the WireMock dependency to your `pom.xml` file.
+
++
+--
+.WireMock dependency
+[source,xml]
+----
+<dependency>
+    <groupId>com.github.tomakehurst</groupId>
+    <artifactId>wiremock-jre8</artifactId>
+    <version>{wiremock.version}</version>
+    <scope>test</scope>
+</dependency>
+----
+
+[NOTE]
+====
+Replace `{wiremock.version}` with the version of WireMock you want to use.
+====
+--
+
+. Choose when you want to start your WireMock server. There are two options:
+
+* Start WireMock server as a `QuarkusTestResource`. This means that your WireMock server will be started once before all your tests and stopped after all your tests. This is the option that fits most cases.
+* Start a WireMock server independently for each test. This is useful if you need different behavior for each test.
+
+. Define how the WireMock server should be started.
+
++
+--
+.Starting a WireMock server example
+[source,java]
+----
+WireMockConfiguration config = WireMockConfiguration.wireMockConfig().dynamicPort(); <1>
+WireMockServer wireMockServer = new WireMockServer(config); <2>
+wireMockServer.start(); <3>
+----
+<1> Creates the configuration for the WireMock server. It's a good practice to use a dynamic port when possible. This way, you prevent the tests from failing due to port conflicts. When using a dynamic port is not possible, you can use the `WireMockConfiguration#port(int)` method to use a fixed port number.
+<2> Creates the WireMock server instance.
+<3> Starts the WireMock server.
+--
+
+. Mock the endpoints.
+
++
+--
+.Mocking endpoints example
+[source,java]
+----
+ObjectMapper objectMapper = new ObjectMapper();
+JsonNode greecePayload = objectMapper.readTree(getClass().getResourceAsStream("/country_mock.json")); <1>
+
+wireMockServer.stubFor(WireMock.get(WireMock.urlEqualTo("/rest/v2/name/Greece")) <2>
+                               .willReturn(WireMock.aResponse() <3>
+                                                   .withStatus(200) <4>
+                                                   .withHeader("Content-Type", "application/json") <5>
+                                                   .withJsonBody(greecePayload))); <6>
+----
+<1> Creates the JSON payload that will be returned by the WireMock server. Alternatively, you can use the `ResponseDefinitionBuilder#withBody` method to define a `String` or a `byte[]` as the response body.
+<2> Defines a stub for the `/rest/v2/name/Greece` endpoint.
+<3> Defines the response for the stub.
+<4> Defines the response status.
+<5> Defines the response headers.
+<6> Defines the response body.
+--
+
+. Stop the WireMock server.
+
++
+--
+.Stopping a WireMock server example
+[source,java]
+----
+wireMockServer.stop();
```

Comment:
Maybe it is better to add where exactly and how to call the close method, maybe on the method annotated with @AfterAallw?

Reply:
This is done in the sections below.